sdk/nodejs/resource.ts

```diff
@@ -782,7 +788,30 @@ export class ComponentResource extends Resource {
         // not correspond to a real piece of cloud infrastructure.  As such, changes to it *itself*
         // do not have any effect on the cloud side of things at all.
         super(type, name, /*custom:*/ false, /*props:*/ {}, opts);
-        this.__pulumiComponentResource = true;
+        this.__data = this.initializeAndRegisterOutputs();
+    }
+
+    private async initializeAndRegisterOutputs() {
+        const data = await this.initialize();
+        this.registerOutputs();
+        return data;
+    }
+
+    /**
+     * Can be overridden by a subclass to asynchronously initialize data for this Component
+     * automatically when constructed.  The data will be available immediately for subclass
+     * constructors to use.  To access the data use `.getData`.
+     */
+    protected async initialize(): Promise<TData> {
+        return <TData>undefined!;
+    }
+
+    /**
+     * Retrieves the data produces by [initialize].  The data is immediately available in a
+     * derived class's constructor after the `super(...)` call to `ComponentResource`.
+     */
+    protected getData(): Promise<TData> {
```

Comment:
Can you add some coverage for this to the tests?

Reply:
sure.